### DISCUSSION TITLE
How Ant Media Server can generate periodic previews(snapshots) of the incoming streams on the fly ?

### DISCUSSION THREAD

**Discussion by switchaphon** [ant-media/Ant-Media-Server#5646] [FAQs] [answered]
Is anyone ever user ANtMedia to generate a snapshot of the stream ?

I follow this article https://github.com/ant-media/Ant-Media-Server/wiki/Preview-Guide.  I already setting and enable following parameter:

```
settings.encoderSettingsString=[{"videoBitrate"\:800000,"forceEncode"\:true,"audioBitrate"\:64000,"height"\:360}]
settings.previewGenerate=true
settings.previewOverwrite=false
settings.createPreviewPeriod=5000
settings.addDateTimeToMp4FileName=true
```

I got the snapshot created as a .png under /previews directory. However I face an issue which I can not understand by googling,  We would like to consult you 3 questions:

1. What is the trigger to take a snapshot? Because we found that some stream IDs are generated as periods that define (5000ms) while some stream IDs do not even when they are broadcasting.

The current time in the picture below is 11.33AM but the highlight file is still 7AM or some of them were  yesterday.
![Screenshot 2566-10-04 at 11 33 35](https://github.com/ant-media/Ant-Media-Server/assets/25098267/724b26b5-5136-4b00-8fa8-29a18da71ddf)

2. Why the snapshot file is overwritten even settings.previewOverwrite is set to FALSE

3. How to append date/time as a suffix of each file name? Because the referred document mention to  adds date/time value to preview file names by configure settings.addDateTimeToMp4FileName to TRUE

Thank you for any suggestion.

**Comment by yashtandon113** [comment, marked as answer]
Hi @switchaphon 
Thank you for posing these questions. Let me try to answer.

1. If the preview is enabled, a preview will be generated for each new stream Id that is published after the setting is applied. So, if certain streams were publishing before the adjustments were applied, those streams should be restarted.

It's possible that this is one of the reasons you don't get previews of some older running streams. I tested it locally on my end with AMS v2.6.4 and it appears to be alright.

Please let us know if the problem persists even after restarting all streams.

2. When the stream is terminated, the most recent preview is kept with the finished suffix, as you can see in below output.

yash@LAPTOP-JRPPOOTM:/usr/local/antmedia/webapps/WebRTCAppEE/previews$ ls
**test.png**  test1_**finished**-2023-10-05_01-56-36.561.png  test_**finished**-2023-10-05_01-53-48.499.png

When the same stream is published again, it will generate a new preview (with a stream Id name like test.png) rather than overwriting the existing one.

3. As I can see in your screenshot, the date and time have been included to the previews. Similarly, you can see in my output.

Please let me know if any of my answers are unclear or if I have missed any points.

**Comment by switchaphon** [comment]
Thank you @yashtandon113 for your prompt response.

1)  Maybe as your suggestion, I deleted all previews file under /previews then stop broadcasting. Re configure an APPNAME then start broadcasting again. The all streams are captured previews as they should do.

**09:56AM**
![Screen Shot 2566-10-05 at 09 56 51](https://github.com/ant-media/Ant-Media-Server/assets/25098267/bf7f712d-c9ed-4b86-adcf-59fc2ab88662)

**10:00AM**
![Screen Shot 2566-10-05 at 10 00 50](https://github.com/ant-media/Ant-Media-Server/assets/25098267/981b2775-4a87-4659-b0a8-f2b64cb9662c)

**10:07AM**
![Screen Shot 2566-10-05 at 10 07 22](https://github.com/ant-media/Ant-Media-Server/assets/25098267/fe2de0c4-ddf8-4838-9fb5-caeacc7e80bc)

2. Dose "stream is terminated" equivalent "stopping a broadcast" through AntMedia UI as well as CCTV itself turn-off ?

3. As your information that mean the suffix with **_YYYY-MM-DD_hh-mm-ss.sss_** will be added only when it trigger by one of the action as 2) ? IRefer to the document I though the every captured preview will be prepend by **__N_** every time AntMedia capturing when  _settings.previewOverwrite=false_ is configured

Additional, I have further 2 more questions

4. Is there any solution we can utilize the preview capture feature without enabling "Adaptive bitrate" option since It consume a lot CPU and Memory of AntMedia Server (TT)

![Screen Shot 2566-10-05 at 10 03 42](https://github.com/ant-media/Ant-Media-Server/assets/25098267/fa491b2c-aad2-4524-9266-f3f8abb830e3)

AntMedia (**_Enterprise Edition 2.5.3 20230107_0851_**) running with _**16vCPU and 32GB of memory**_ where only **_APPNAME2_** (w/ 3 Streams) and **_APPNAME3_** (w/ 16 streams) which enable an "Adaptive Bitrate" function with 360p and 480p

![Screen Shot 2566-10-05 at 10 20 26](https://github.com/ant-media/Ant-Media-Server/assets/25098267/bb042af0-4acb-48d4-bf5e-f2e95d925621)

![Screen Shot 2566-10-05 at 10 20 32](https://github.com/ant-media/Ant-Media-Server/assets/25098267/c177e84d-a9b8-4d96-a669-c86ecb9caeea)

5. Is there any API to support this preview capture which we can manual request instead ?


Again Thank you! for your kind support. :)

**Reply by yashtandon113** [reply]
@switchaphon 
Let me answer your other questions.

2. Yes, stopping the stream from the server and stopping it from the source are the same thing.

3. You are correct. Because the preview is generated every 5000ms, a date and time suffix will be appended to it after the stream is stopped. In the end, the date and time will be added so that they are not overwritten when the stream is published again.

4. Currently, there is no method to generate the preview without enabling ABR on the server side. You are asked to submit a feature request [here](https://github.com/ant-media/Ant-Media-Server/issues/new?assignees=&labels=&projects=&template=feature_request.md&title=).

5. Unfortunately, there is no API to capture the preview on the fly.

If you are not recording the streams, you can record a tiny mp4 clip on the server using the start/stop recording API call and then get the one frame using FFMPEG and use it as a preview. The mp4 file can then be removed after that.

I hope it helps.
